### PR TITLE
Adding Eigen::Ref to H5Easy (and simplifying Eigen overload)

### LINKED ISSUE #286: Writing to hdf: "static assertion failed" with Eigen matrix, mapped from OpenCV
* HighFive => 2.1-dev (current master)
* Eigen => 3.3.7
* OpenCV => 4.1.2, 4.2.0-dev
* Operating System / Platform => Debian 10 x86_64 GNU/Linux (4.19.0-6-amd64)
* Compiler => g++ 8.3.0

##### Detailed description

I can do an `Eigen::Map` of a `cv::Mat`, to recover the data in the form of an `Eigen::Matrix`, but I get a 'static assertion failed' as soon as I use HighFive to store that data in an hdf file.

Since the mapping appears to work, I do not understand why `H5Easy::dump` behaves differently than with any another `Eigen::Matrix`.

###### 1. Eigen::Map:
```cpp

double data[16] = {1,2,3,4, 5,6,7,8, 9,10,11,12, 13,14,15,16};
cv::Mat cvmat = cv::Mat_<double>(4,4, data);

// fixed dimensions, column-major
Eigen::Map<Eigen::Matrix4d> eigen_simple(reinterpret_cast<double*>(cvmat.data));
```

###### 2. Behaves as expected:
```cpp
std::cout << eigen_simple << '\n' << std::endl;
std::cout << eigen_simple.row(0) << '\n' << std::endl;
std::cout << eigen_simple.col(0) << '\n' << std::endl;
```
indeed prints
```
 1  5  9 13
 2  6 10 14
 3  7 11 15
 4  8 12 16

 1  5  9 13

1
2
3
4
```

###### 3. Won't compile once I try to write the data to an hdf file:
```cpp
H5Easy::File output_hdf("out_test.hdf", H5Easy::File::Overwrite);

// won't compile:
H5Easy::dump(output_hdf, "/Eigen_simple", eigen_simple, H5Easy::DumpMode::Overwrite);
```
Here's the error:
```
/usr/local/include/highfive/bits/H5Converter_misc.hpp: In instantiation of ‘HighFive::details::data_converter<Scalar, Enable>::data_converter(Scalar&, HighFive::DataSpace&) [with Scalar = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Enable = void]’:
/usr/local/include/highfive/bits/H5Slice_traits_misc.hpp:227:44:   required from ‘void HighFive::SliceTraits< <template-parameter-1-1> >::write(const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Derivate = HighFive::DataSet]’
/usr/local/include/highfive/h5easy_bits/H5Easy_scalar.hpp:29:9:   required from ‘static HighFive::DataSet H5Easy::detail::scalar::dump_impl<T, E>::run(HighFive::File&, const string&, const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; E = void; std::__cxx11::string = std::__cxx11::basic_string<char>]’
/usr/local/include/highfive/h5easy_bits/H5Easy_scalar.hpp:137:49:   required from ‘HighFive::DataSet H5Easy::dump(HighFive::File&, const string&, const T&, H5Easy::DumpMode) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/file/main.cpp:39:85:   required from here
/usr/local/include/highfive/bits/H5Converter_misc.hpp:121:52: error: static assertion failed: supported datatype should be an arithmetic value, a std::string or a container/array
         static_assert((std::is_arithmetic<Scalar>::value ||
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                        std::is_enum<Scalar>::value ||
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
                        std::is_same<std::string, Scalar>::value),
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The message is the same as in #256, but note that I am *not using Boost* here (so #261 won't help); I actually opted out when compiling HighFive. Note also that the problem remains with the current master (which already includes patch #258).

##### Full example

```cpp
#include <iostream>
#include <Eigen/Eigen>
#include <opencv2/core/eigen.hpp>
#include <highfive/H5Easy.hpp>

namespace eigentools {
	template<class T>
	using MatrixX = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>; 
}


int main()
{

	double data[16] = {1,2,3,4, 5,6,7,8, 9,10,11,12, 13,14,15,16};
	cv::Mat cvmat = cv::Mat_<double>(4,4, data);

	// fixed dimensions, column-major
	Eigen::Map<Eigen::Matrix4d> eigen_simple(reinterpret_cast<double*>(cvmat.data));
	std::cout << eigen_simple << '\n' << std::endl;

	// dynamic, row-major
	Eigen::Map<eigentools::MatrixX<double>> eigen_dynamic_rowmajor(cvmat.ptr<double>(), cvmat.rows, cvmat.cols); 
	std::cout << eigen_dynamic_rowmajor << '\n' << std::endl;

	// idem
	Eigen::Map<eigentools::MatrixX<double>> eigen_dynamic_rowmajor_alt(reinterpret_cast<double*>(cvmat.data), cvmat.rows, cvmat.cols);
	std::cout << eigen_dynamic_rowmajor_alt << '\n' << std::endl;

	H5Easy::File output_hdf("out_test.hdf", H5Easy::File::Overwrite);

	// So far, so good.

	// However, putting any of these will prevent the programme from compiling:

	H5Easy::dump(output_hdf, "/Eigen_simple", eigen_simple, H5Easy::DumpMode::Overwrite);
	H5Easy::dump(output_hdf, "/EigenRowMajorFromCV", eigen_dynamic_rowmajor, H5Easy::DumpMode::Overwrite);
	H5Easy::dump(output_hdf, "/EigenRowMajorFromCV_alt", eigen_dynamic_rowmajor_alt, H5Easy::DumpMode::Overwrite);

	return 0;
}
```

### PR DESCRIPTION
**Pull request by tdegeus**
Fixes #286 

It needed to be verified if the CRTP is satisfied. In any case the implementation offers a significant simplification in the internal implementation.

### CODE REVIEW

**Comment by spacescientist** [comment]
Thank you for this, but I am a bit confused: did it actually work with `Eigen::Map` on your side? 

I get these errors (the first one is new).

```
In file included from path/to/lib/include/highfive/H5DataSpace.hpp:158,
                 from path/to/lib/include/highfive/H5Attribute.hpp:14,
                 from path/to/lib/include/highfive/H5File.hpp:75,
                 from path/to/lib/include/highfive/H5Easy.hpp:38,
                 from path/to/project/MWE_using_highfivemaster/main.cpp:5:
path/to/lib/include/highfive/bits/H5Dataspace_misc.hpp: In instantiation of ‘static HighFive::DataSpace HighFive::DataSpace::From(const ScalarValue&) [with ScalarValue = Eigen::Map<Eigen::Matrix<double, 4, 4> >]’:
path/to/lib/include/highfive/h5easy_bits/H5Easy_scalar.hpp:28:70:   required from ‘static HighFive::DataSet H5Easy::detail::scalar::dump_impl<T, E>::run(HighFive::File&, const string&, const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; E = void; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/lib/include/highfive/h5easy_bits/H5Easy_scalar.hpp:137:49:   required from ‘HighFive::DataSet H5Easy::dump(HighFive::File&, const string&, const T&, H5Easy::DumpMode) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/project/MWE_using_highfivemaster/main.cpp:41:85:   required from here
path/to/lib/include/highfive/bits/H5Dataspace_misc.hpp:136:43: error: static assertion failed: Only the following types are supported by DataSpace::From: 
  signed_arithmetic_types = int | long | float | double 
  unsigned_arithmetic_types = unsigned signed_arithmetic_types 
  string_types = std::string 
  all_basic_types = string_types | unsigned_arithmetic_types | signed_arithmetic_types 
   stl_container_types = std::vector<all_basic_types>   boost_container_types = boost::numeric::ublas::matrix<all_basic_types> | boost::multi_array<all_basic_types> 
  all_supported_types = all_basic_types | stl_container_types | boost_container_types  eigen_matrix_type = Eigen::Matrix<signed_arithmetic_type> | Eigen::VectorXX
         (std::is_arithmetic<ScalarValue>::value ||
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          std::is_enum<ScalarValue>::value ||
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
          std::is_same<std::string, ScalarValue>::value),
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Followed by what very much looks like the original error:
```
In file included from path/to/lib/include/highfive/bits/H5Attribute_misc.hpp:25,
                 from path/to/lib/include/highfive/H5Attribute.hpp:75,
                 from path/to/lib/include/highfive/H5File.hpp:75,
                 from path/to/lib/include/highfive/H5Easy.hpp:38,
                 from path/to/project/MWE_using_highfivemaster/main.cpp:5:
path/to/lib/include/highfive/bits/H5Converter_misc.hpp: In instantiation of ‘HighFive::details::data_converter<Scalar, Enable>::data_converter(Scalar&, HighFive::DataSpace&) [with Scalar = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Enable = void]’:
path/to/lib/include/highfive/bits/../bits/H5Slice_traits_misc.hpp:246:44:   required from ‘void HighFive::SliceTraits< <template-parameter-1-1> >::write(const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; Derivate = HighFive::DataSet]’
path/to/lib/include/highfive/h5easy_bits/H5Easy_scalar.hpp:29:9:   required from ‘static HighFive::DataSet H5Easy::detail::scalar::dump_impl<T, E>::run(HighFive::File&, const string&, const T&) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; E = void; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/lib/include/highfive/h5easy_bits/H5Easy_scalar.hpp:137:49:   required from ‘HighFive::DataSet H5Easy::dump(HighFive::File&, const string&, const T&, H5Easy::DumpMode) [with T = Eigen::Map<Eigen::Matrix<double, 4, 4> >; std::__cxx11::string = std::__cxx11::basic_string<char>]’
path/to/project/MWE_using_highfivemaster/main.cpp:41:85:   required from here
path/to/lib/include/highfive/bits/H5Converter_misc.hpp:121:52: error: static assertion failed: supported datatype should be an arithmetic value, a std::string or a container/array
         static_assert((std::is_arithmetic<Scalar>::value ||
                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                        std::is_enum<Scalar>::value ||
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
                        std::is_same<std::string, Scalar>::value),
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/executable.dir/build.make:63: CMakeFiles/executable.dir/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/executable.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

**Comment by tdegeus** [comment]
No it seems to match the wrong overload. Do you have a minimal example for me to try locally?

**Comment by spacescientist** [comment]
Alright, I was wondering whether I might have missed something. 

So I wrote a MWE with `std::vector` in https://github.com/BlueBrain/HighFive/issues/286#issuecomment-588893348. Cheers.

**Comment by tdegeus** [comment]
It works now. The only subtlety was/is to convince the compiler to select the Eigen overload, so that's why I added the `Eigen::Map` and `Eigen::Ref` overloads that are a perfect match.

**Comment by tdegeus** [comment]
Ready for review!

**Comment by alexsavulescu** [comment]
@tdegeus could you also add the Eigen::Map MWE in the unit test?

**Comment by tdegeus** [comment]
Done @alexsavulescu !

File: include/highfive/H5Easy.hpp [hunk 1]
```diff
@@ -91,6 +91,24 @@ inline DataSet dump(File& file,
...
+/// \param file Writeable opened file
+/// \param path Path of the DataSet
+/// \param data eigen matrix to write
+/// \param mode Write mode
+///
+/// \return the newly created DataSet
+///
+#ifdef H5_USE_EIGEN
+template <class T>
+inline DataSet dump(File& file,
```
Comment (eudoxos) [inline include/highfive/H5Easy.hpp]: I *think* using `template DerivedT` inline DataSet dump(..., Eigen::DenseBase<DerivedT>){...}` the dump could be resolved for many more Eigen types (Arrays, Matrices, References, expression...). Maybe beyond the scope of the this PR, though.

Comment (tdegeus) [reply include/highfive/H5Easy.hpp]: In principle you are right. However, what complicates matters here is that there is also a 
```cpp
template <class T>
inline DataSet dump(File& file,
                    const std::string& path,
                    const T& data,
                    DumpMode mode = DumpMode::Create);
```
that takes all HighFive's know types (`int`, `double`, ..., `std::vector<T>`). The trick is to convince the compiler not to take it when calling with an Eigen object. 

My approach is to include overloads of the actual objects that call `dump` (and not their parents). What could be more elegant is to have a static assert in the overload I mentioned. But I'm not entirely sure how to play that out.

Comment (alexsavulescu) [reply include/highfive/H5Easy.hpp]: Thank you @eudoxos for your input. We will definitely improve on our Eigen support and will take this into consideration.

Comment (eudoxos) [reply include/highfive/H5Easy.hpp:106]: > The trick is to convince the compiler not to take it when calling with an Eigen object.

Even though this is closed, just a note for the future: `std::is_base_of` might help here, perhaps.

Comment (tdegeus) [reply include/highfive/H5Easy.hpp:106]: That would be a good idea @eudoxos ! Evidently I did not search long enough. Would you be willing to give it a try and open a PR?

**Pull request merged**